Add: Support for Sparse24Bitmask Compressed Models

rahul-tuli opened the pull request:
This PR adds support for models compressed using `Sparse24BitMaskCompressor` to use cutlass 2:4 Kernels

- Adds support for compressed cases

This diff was manually tested on:

- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-chnl_wts_per_tok_dyn_act_fp8-BitM`
- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-chnl_wts_per_tok_dyn_act_int8-BitM`
- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-tensor_wts_tensor_act_fp8-BitM`
- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-tensor_wts_tensor_act_int8-BitM`
- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-tensor_wts_per_tok_dyn_act_fp8-BitM`
- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-tensor_wts_per_tok_dyn_act_int8-BitM`
- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-chnl_wts_tensor_act_fp8-BitM`
- `nm-testing/SparseLlama-3.1-8B-gsm8k-pruned.2of4-chnl_wts_tensor_act_int8-BitM`

Also added unit tests for the compressed 2:4 fp8, int8, and sparse only cases!!
Notion Doc: https://www.notion.so/SparseBitMask-24-work-15e863ebf65c80dcbc70e6317d552987

rahul-tuli committed:
Add: Support for Sparse24Bitmask Compressed Models

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
Fix: mypy errors

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli commented:
Add a test file with an 8B 2of4 compressed model for lm_eval_harness in buildkite
Add test cases for:
```
-> Sparse only
-> fp8 + sparse dynamic per token
-> fp8 scheme
-> int8 dynamic
-> int8 scheme
```

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:
```diff
@@ -481,6 +495,19 @@ def supports_cutlass_24(
 
         return weight_quant.num_bits == input_quant.num_bits == 8
 
+    def _get_model_compression_config(
```
seems like an unnecessary function break out

rahul-tuli replied:
Removed

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:
```diff
@@ -57,6 +74,34 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
+            assert all(
+                partition_size % 8 == 0
+                for partition_size in output_partition_sizes
+            ), "All partitions must be divisible by 8 for 2:4 compressed models"
```
maybe "for a 2:4 sparse compressed model"?

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:
```diff
@@ -57,6 +74,34 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
+            shape = BitMaskShapeParameter(data=torch.empty(
+                2 * len(output_partition_sizes), 1, dtype=torch.uint64),
+                                          weight_loader=weight_loader)
+            compressed = ModelWeightParameter(data=torch.empty(
```
nit: parameter name

rahul-tuli replied:
updated to `compressed_weight`?

dsikka replied:
yes

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:
```diff
@@ -201,8 +253,42 @@ def _get_params_dtype(self, params_dtype: torch.dtype) -> torch.dtype:
-    new_tensor = tensor.view(-1, 4)
-    zero_counts = (new_tensor == 0).sum(dim=1)
-    return (zero_counts >= 2).all().item()
+    def _decompress_bitmask_compressed_weight(
```
docstring

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:
```diff
@@ -57,6 +74,34 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
+                for partition_size in output_partition_sizes
+            ), "All partitions must be divisible by 8 for 2:4 compressed models"
+
+            shape = BitMaskShapeParameter(data=torch.empty(
```
We dont need to shard the shape

rahul-tuli replied:
Removed!

rahul-tuli committed:
Removed BitmaskShape Parameter
Renamed `compressed` to `compressed_weight`
Address review commits from @dsikka

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
Add: lm-eval, fp8, int8 tests

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
Add: 2:4 Sparse only compressed test

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
Lint

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

mgoin commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:
```diff
@@ -112,6 +156,13 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         """
+        if self.do_sparse_decompress:
+            layer.weight.data = self._decompress_bitmask_compressed_weight(
+                compressed=layer.compressed,
+                bitmask=layer.bitmask,
```
Should we delete `layer.compressed` and `layer.bitmask` after decompressing them?

rahul-tuli replied:
We absolutely can, addressed in latest commit!

rahul-tuli committed:
Delete: compression params after decompression

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
Update: Compressed Tensors version

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

dsikka reviewed (commented):
can you bump ct to 0.9.1

tlrmchlsmth commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:
```diff
@@ -201,8 +257,55 @@ def _get_params_dtype(self, params_dtype: torch.dtype) -> torch.dtype:
+        
+        This function also supports sharded decompression.
+
+        :param compressed: The 2:4 sparse weight tensor 
+            compressed using the sparse-24-bitmask compressor.
```
I like the decision to use bitmask compression for 2:4 sparse weights - simple and reuses existing decompression implementations. However I do think that there's a potential source of confusion here just because the 2:4 instructions use a different scheme altogether (2 bits for every nonzero that represent the coordinate within the block of 4). Could you add an explanatory comment pointing out the difference between the two schemes?

tlrmchlsmth commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py:
```diff
@@ -112,6 +156,18 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+                bitmask=layer.bitmask,
+                layer=layer,
+            )
+            # compressed and bitmask tensors
+            # are no longer needed after decompression
+            
+            del layer.compressed
+            del layer.bitmask
```
nit: better whitespace like this?
```suggestion

            # compressed and bitmask tensors
            # are no longer needed after decompression
            del layer.compressed
            del layer.bitmask
```

rahul-tuli committed:
Update vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_24.py

Co-authored-by: Tyler Michael Smith <tyler@neuralmagic.com>
Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
Address: Review comments

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
pre-commit hooks
Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

rahul-tuli committed:
test-file ruff

Signed-off-by: Rahul Tuli <rahul@neuralmagic.com>

mgoin committed:
Merge branch 'main' into rahul-bitmask-additions
